suite/tests/api/ir_rvv.c

```diff
@@ -0,0 +1,1852 @@
+
+    ASSERT(instr_get_opcode(instr) == opcode);
+    instr_disassemble(dc, instr, STDERR);
+    print("\n");
+    ASSERT(instr_is_encoding_possible(instr));
+    pc = instr_encode(dc, instr, buf);
+    ASSERT(pc != NULL);
+    decin = instr_create(dc);
+    next_pc = decode(dc, buf, decin);
+    ASSERT(next_pc != NULL);
+    if (!instr_same(instr, decin)) {
+        print("Disassembled as:\n");
+        instr_disassemble(dc, decin, STDERR);
+        print("\n");
+        ASSERT(instr_same(instr, decin));
+    }
+
+    instr_destroy(dc, instr);
+    instr_destroy(dc, decin);
+    return pc;
+}
+
+static void
+test_instr_encoding_failure(void *dc, uint opcode, app_pc instr_pc, instr_t *instr)
+{
+    byte *pc;
+
+    pc = instr_encode_to_copy(dc, instr, buf, instr_pc);
+    ASSERT(pc == NULL);
+    instr_destroy(dc, instr);
+}
+
+static byte *
+test_instr_decoding_failure(void *dc, uint raw_instr)
+{
+    instr_t *decin;
+    byte *pc;
+
+    *(uint *)buf = raw_instr;
+    decin = instr_create(dc);
+    pc = decode(dc, buf, decin);
+    /* Returns NULL on failure. */
+    ASSERT(pc == NULL);
+    instr_destroy(dc, decin);
+    return pc;
+}
+
+static void
+test_configuration_setting(void *dc)
+{
+    instr_t *instr;
+
+    instr = INSTR_CREATE_vsetivli(dc, opnd_create_reg(DR_REG_A1),
+                                  opnd_create_immed_int(0b01010, OPSZ_5b),
+                                  opnd_create_immed_int(0b00001000, OPSZ_10b));
+    test_instr_encoding(dc, OP_vsetivli, instr);
+    instr =
+        INSTR_CREATE_vsetvli(dc, opnd_create_reg(DR_REG_A1), opnd_create_reg(DR_REG_A2),
+                             opnd_create_immed_int(0b000001000, OPSZ_11b));
+    test_instr_encoding(dc, OP_vsetvli, instr);
+    instr = INSTR_CREATE_vsetvl(dc, opnd_create_reg(DR_REG_A1),
+                                opnd_create_reg(DR_REG_A2), opnd_create_reg(DR_REG_A3));
+    test_instr_encoding(dc, OP_vsetvl, instr);
+}
+
+static void
+test_unit_stride(void *dc)
+{
+    instr_t *instr;
+
+    instr = INSTR_CREATE_vlm_v(
+        dc, opnd_create_reg(DR_REG_VR0),
+        opnd_create_base_disp(DR_REG_A1, DR_REG_NULL, 0, 0, reg_get_size(DR_REG_VR0)));
+    test_instr_encoding(dc, OP_vlm_v, instr);
+    instr = INSTR_CREATE_vsm_v(
+        dc, opnd_create_base_disp(DR_REG_A1, DR_REG_NULL, 0, 0, reg_get_size(DR_REG_VR0)),
+        opnd_create_reg(DR_REG_VR0));
+    test_instr_encoding(dc, OP_vsm_v, instr);
+    instr = INSTR_CREATE_vle8_v(
+        dc, opnd_create_reg(DR_REG_VR0),
+        opnd_create_base_disp(DR_REG_A1, DR_REG_NULL, 0, 0, reg_get_size(DR_REG_VR0)),
+        opnd_create_immed_int(0b1, OPSZ_1b), opnd_create_immed_int(0b000, OPSZ_3b));
+    test_instr_encoding(dc, OP_vle8_v, instr);
+    instr = INSTR_CREATE_vle16_v(
+        dc, opnd_create_reg(DR_REG_VR0),
+        opnd_create_base_disp(DR_REG_A1, DR_REG_NULL, 0, 0, reg_get_size(DR_REG_VR0)),
+        opnd_create_immed_int(0b1, OPSZ_1b), opnd_create_immed_int(0b000, OPSZ_3b));
```

Comment:
There is a lot of duplication here and in all the functions below.  This would all be easier to read and maintain if we shared the details in a central place.  The x86 opcode tests split them up by number and types of args.  Start with simple local macros?  So this would turn into something like `TEST_REG_MEM_IMM_IMM(vle16_v);` and above `TEST_MEM_REG(vsm_v);`.  Then we have just one line per opcode in these tests.

Reply:
I add some local macros in this file and replace all duplication.